core/runtime/impl/block_builder_impl.cpp

```diff
@@ -0,0 +1,139 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "runtime/impl/block_builder_impl.hpp"
+
+#include <utility>
+#include <vector>
+
+#include "runtime/impl/wasm_memory_stream.hpp"
+#include "scale/collection.hpp"
+#include "scale/type_decoder.hpp"
+
+namespace kagome::runtime {
+
+  using common::Buffer;
+  using extensions::Extension;
+  using primitives::Extrinsic;
+  using primitives::ScaleCodec;
+  using scale::collection::decodeCollection;
+  using wasm::Literal;
+
+  BlockBuilderImpl::BlockBuilderImpl(Buffer state_code,
+                                     std::shared_ptr<Extension> extension,
+                                     std::shared_ptr<ScaleCodec> codec)
+      : memory_{extension->memory()},
+        codec_{std::move(codec)},
+        executor_{std::move(extension)},
+        state_code_{std::move(state_code)} {}
+
+  outcome::result<bool> BlockBuilderImpl::apply_extrinsic(Extrinsic extrinsic) {
+    OUTCOME_TRY(encoded_ext, codec_->encodeExtrinsic(extrinsic));
+
+    runtime::SizeType ext_size = encoded_ext.size();
+    runtime::WasmPointer ptr = memory_->allocate(ext_size);
```

Comment:
`Memory::allocate(..)` can return non-success values, such as 0 or -1. Please, check for them